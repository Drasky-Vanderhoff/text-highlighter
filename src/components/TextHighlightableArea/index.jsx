import React, {useState, useRef, Fragment} from 'react'
import PropTypes from 'prop-types'
import './index.scss'

const TextHighlightableArea = ({initialText, onChangeText, onSelectedText, highlightMarkers}) => {
  const [text, setText] = useState(initialText)
  const textareaEl = useRef(null)

  /* Callbacks */
  const handleSelection = () => {
    const {selectionStart, selectionEnd} = textareaEl.current
    if (selectionStart !== selectionEnd) onSelectedText(selectionStart, selectionEnd)
  }
  
  const handleInput = (event) => {
    const newText = event.target.value
    onChangeText(newText)
    setText(newText)
  }
  
  /* Renders */

  /* Renders the text previous to the marked text and them the marker, 
   * the rest is being handle in posterior interations that will keep consuming the buffer.
   * Marker are consider to be consider already ordered ( this is being done by the reducer ).
   */
  const renderHighlights = (input, highlightMarkers) => {
    let offset = 0
    let buffer = input
    return highlightMarkers.length ? highlightMarkers.map(
      ({selectionStart, selectionEnd, color}, idx) => {
        const result = (
          <Fragment key={idx}>
            {buffer.slice(0, selectionStart - offset)}
            <mark style={{backgroundColor: color}}>{
              buffer.slice(selectionStart - offset, selectionEnd - offset
            )}</mark>
          </Fragment>
        )
        buffer = buffer.slice(selectionEnd - offset)
        offset = selectionEnd
        return result
      }        
    ) : buffer
  }

  return (
    <div className="text-highlightable-area">
      <div className="backdrop">
        <div 
          className="highlights"
          data-testid="highlights"
        >{
          renderHighlights(text, highlightMarkers)
        }</div>
      </div>
      <textarea
        data-testid="text-area"
        ref={textareaEl}
        value={text}
        onTouchEnd={handleSelection} 
        onClick={handleSelection}
        onChange={handleInput}
      />
    </div>
  )
}

export const TextHighlightableAreaPropTypes = {
  initialText: PropTypes.string,
  onChangeText: PropTypes.func,
  onSelectedText: PropTypes.func,
  highlightMarkers: PropTypes.arrayOf(PropTypes.shape({
    selectionStart: PropTypes.number,
    selectionEnd: PropTypes.number,
    color: PropTypes.string
  }))
}

TextHighlightableArea.propTypes = TextHighlightableAreaPropTypes

TextHighlightableArea.defaultProps = {
  initialText: "",
  onChangeText: () => {},
  onSelectedText: () => {},
  highlightMarkers: []
}

export default TextHighlightableArea