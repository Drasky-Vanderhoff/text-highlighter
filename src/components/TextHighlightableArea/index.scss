.text-highlightable-area {
  display: block;
  width: 80%;
  height: 300px;
  margin-left: 5%;
  transform: translateZ(0);
  -webkit-text-size-adjust: none;

  *, *::before, *::after {
    box-sizing: border-box;
  }
  
  .backdrop, textarea {
    width: 100%;
    height: 100%;
  }

  .highlights, textarea {
    padding: 10px;
    font: 20px/28px 'Open Sans', sans-serif;
    letter-spacing: 1px;
  }

  .backdrop {
    position: absolute;
    z-index: 1;
    border: none;
    background-color: #fff;
    overflow: auto;
    pointer-events: none;
    transition: transform 1s;
  }

  .highlights {
    white-space: pre-wrap;
    word-wrap: break-word;
    color: transparent;
  }

  textarea {
    display: block;
    position: absolute;
    z-index: 2;
    border: none;
    color: #444;
    background-color: transparent;
    overflow: hidden;
    text-overflow: ellipsis;
    resize: none;
    transition: transform 1s;
    -ms-overflow-style: none;
  }
}