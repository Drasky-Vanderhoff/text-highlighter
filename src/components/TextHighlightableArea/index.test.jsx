import React from 'react'
import {render, fireEvent} from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import TextHighlightableArea from './'
import renderer from 'react-test-renderer'

describe('TextHighlightableArea Component', () => {
  test('should render with default values', () => {
    expect(
      renderer.create(<TextHighlightableArea />).toJSON()
    ).toMatchSnapshot()
  })

  test('should render with text highlighted', () => {
    const props = {
      initialText: "Lorem Ipsum is simply dummy text of the printing and typesetting industry.",
      highlightMarkers: [
        {
          selectionStart: 0,
          selectionEnd: 5,
          color: "red"
        },
        {
          selectionStart: 12,
          selectionEnd: 21,
          color: "blue"
        }
      ]
    }
    expect(
      renderer.create(<TextHighlightableArea {...props}/>).toJSON()
    ).toMatchSnapshot()
  })

  test('should call change text callback handler when text is changed by the user in the text area', () => {
    const props = {
      initialText: "Lorem Ipsum is simply dummy text of the printing and typesetting industry.",
      onChangeText: jest.fn()
    }
    const newText = "This is the new text"
    const {getByTestId} = render(<TextHighlightableArea {...props} />)
    /* Precondition */
    expect(getByTestId("text-area")).toHaveTextContent(props.initialText) 
    expect(getByTestId("highlights")).toHaveTextContent(props.initialText)
    expect(props.onChangeText).toHaveBeenCalledTimes(0)

    /* Action */
    fireEvent.change(getByTestId("text-area"), {target: { value: newText}})

    /* Postcondition */
    expect(getByTestId("text-area")).toHaveTextContent(newText)
    expect(getByTestId("highlights")).toHaveTextContent(newText)
    expect(props.onChangeText).toHaveBeenCalledTimes(1)
    expect(props.onChangeText).toHaveBeenCalledWith(newText)
  })

  test('should call change text callback handler when text is changed by the user in the text area and update the highlight area', () => {
    const props = {
      initialText: "Lorem Ipsum is simply dummy text of the printing and typesetting industry.",
      onChangeText: jest.fn()
    }
    const newText = "This is the new text"
    const {getByTestId} = render(<TextHighlightableArea {...props} />)
    /* Precondition */
    expect(getByTestId("text-area")).toHaveTextContent(props.initialText) 
    expect(getByTestId("highlights")).toHaveTextContent(props.initialText)
    expect(props.onChangeText).toHaveBeenCalledTimes(0)

    /* Action */
    fireEvent.change(getByTestId("text-area"), {target: { value: newText}})

    /* Postcondition */
    expect(getByTestId("text-area")).toHaveTextContent(newText)
    expect(getByTestId("highlights")).toHaveTextContent(newText)
    expect(props.onChangeText).toHaveBeenCalledTimes(1)
    expect(props.onChangeText).toHaveBeenCalledWith(newText)
  })
  
  test('should call selection text callback handler when text is selected in desktop environment', () => {
    const props = {
      initialText: "Lorem Ipsum is simply dummy text of the printing and typesetting industry.",
      onSelectedText: jest.fn()
    }
    const [selectionStart, selectionEnd] = [0, 5]
    
    const {getByTestId} = render(<TextHighlightableArea {...props} />)
    /* Precondition */
    expect(getByTestId("text-area")).toHaveTextContent(props.initialText) 
    expect(getByTestId("highlights")).toHaveTextContent(props.initialText)
    expect(props.onSelectedText).toHaveBeenCalledTimes(0)
    
    /* Action */
    fireEvent.click(getByTestId("text-area"), {target: { selectionStart, selectionEnd}})

    /* Postcondition */
    expect(props.onSelectedText).toHaveBeenCalledTimes(1)
    expect(props.onSelectedText).toHaveBeenCalledWith(selectionStart, selectionEnd)
  })  
  
  test('should not call selection text callback handler when text selection is empty in desktop environment', () => {
    const props = {
      initialText: "Lorem Ipsum is simply dummy text of the printing and typesetting industry.",
      onSelectedText: jest.fn()
    }
    const [selectionStart, selectionEnd] = [5, 5]
    
    const {getByTestId} = render(<TextHighlightableArea {...props} />)
    /* Precondition */
    expect(getByTestId("text-area")).toHaveTextContent(props.initialText) 
    expect(getByTestId("highlights")).toHaveTextContent(props.initialText)
    expect(props.onSelectedText).toHaveBeenCalledTimes(0)
    
    /* Action */
    fireEvent.click(getByTestId("text-area"), {target: { selectionStart, selectionEnd}})

    /* Postcondition */
    expect(props.onSelectedText).toHaveBeenCalledTimes(0)
  })
  
  test('should call selection text callback handler when text is selected in mobile environment', () => {
    const props = {
      initialText: "Lorem Ipsum is simply dummy text of the printing and typesetting industry.",
      onSelectedText: jest.fn()
    }
    const [selectionStart, selectionEnd] = [0, 5]
    
    const {getByTestId} = render(<TextHighlightableArea {...props} />)
    /* Precondition */
    expect(getByTestId("text-area")).toHaveTextContent(props.initialText) 
    expect(getByTestId("highlights")).toHaveTextContent(props.initialText)
    expect(props.onSelectedText).toHaveBeenCalledTimes(0)
    
    /* Action */
    fireEvent.touchEnd(getByTestId("text-area"), {target: { selectionStart, selectionEnd}})

    /* Postcondition */
    expect(props.onSelectedText).toHaveBeenCalledTimes(1)
    expect(props.onSelectedText).toHaveBeenCalledWith(selectionStart, selectionEnd)
  })
  
  test('should not call selection text callback handler when text selection is empty in mobile environment', () => {
    const props = {
      initialText: "Lorem Ipsum is simply dummy text of the printing and typesetting industry.",
      onSelectedText: jest.fn()
    }
    const [selectionStart, selectionEnd] = [5, 5]
    
    const {getByTestId} = render(<TextHighlightableArea {...props} />)
    /* Precondition */
    expect(getByTestId("text-area")).toHaveTextContent(props.initialText) 
    expect(getByTestId("highlights")).toHaveTextContent(props.initialText)
    expect(props.onSelectedText).toHaveBeenCalledTimes(0)
    
    /* Action */
    fireEvent.touchEnd(getByTestId("text-area"), {target: { selectionStart, selectionEnd}})

    /* Postcondition */
    expect(props.onSelectedText).toHaveBeenCalledTimes(0)
  })
  
})
