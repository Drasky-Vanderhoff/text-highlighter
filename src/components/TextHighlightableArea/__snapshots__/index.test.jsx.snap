// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextHighlightableArea Component should render with default values 1`] = `
<div
  className="text-highlightable-area"
>
  <div
    className="backdrop"
  >
    <div
      className="highlights"
      data-testid="highlights"
    >
      
    </div>
  </div>
  <textarea
    data-testid="text-area"
    onChange={[Function]}
    onClick={[Function]}
    onTouchEnd={[Function]}
    value=""
  />
</div>
`;

exports[`TextHighlightableArea Component should render with text highlighted 1`] = `
<div
  className="text-highlightable-area"
>
  <div
    className="backdrop"
  >
    <div
      className="highlights"
      data-testid="highlights"
    >
      
      <mark
        style={
          Object {
            "backgroundColor": "red",
          }
        }
      >
        Lorem
      </mark>
       Ipsum 
      <mark
        style={
          Object {
            "backgroundColor": "blue",
          }
        }
      >
        is simply
      </mark>
    </div>
  </div>
  <textarea
    data-testid="text-area"
    onChange={[Function]}
    onClick={[Function]}
    onTouchEnd={[Function]}
    value="Lorem Ipsum is simply dummy text of the printing and typesetting industry."
  />
</div>
`;
